controllers/ideaController.js

```diff
@@ -12,77 +20,185 @@ exports.createIdea = function(req,res){
 			res.send(util.wrapBody({success:true}));
 		}
 	});
-}
+
+};
 
 exports.deleteIdea = function(req,res){
-	var ideaId = req.params.id;
+	
 
-	util.checkParams(req.params,['id'],function(err){
+	util.checkParam(req.params,['id'],function(err){
 		if (err) {
 			console.log(err);
-			res.send(util.wrapBody('Internal Err','E'));
+			res.send(util.wrapBody('Invalid Parameter','E'));
 		} else {
+			var ideaId = req.params.id;
+
 			ideaRepostory.deleteById(ideaId,function(err){
 				if (err) {
 					console.log(err);
 					res.send(util.wrapBody('Internal Err','E'));
 				} else {
 					res.send(util.wrapBody({success:true}));
 				}
-			})
+			});
 		}
-	})
-
-}
+	});
+};
 
 exports.listIdea = function(req,res){
-	util.checkParams(req.body,['pageSize','pageNum'],function(err){
+	var conditions = req.body;
+
+	ideaRepostory.query(conditions,function(err,result){
 		if (err) {
 			console.log(err);
-			res.send(util.wrapBody('Internal Err','E'));
+			res.send(util.wrapBody('Internal Error','E'));
 		} else {
-			var pageSize = req.body.pageSize;
-			var pageNum = req.body.pageNum;
+			res.send(util.wrapBody(result));
+		}
+	});
 
-			var options = {
-				pageNum:pageNum,
-				pageSize:pageSize
-			}
+};
 
-			ideaRepostory.queryIdea(options,function(err,result){
-				if (err) {
-					console.log(err);
-					res.send(util.wrapBody('Internal Err','E'));
-				} else {
-					res.send(util.wrapBody(result));
+exports.publishIdea = function(req,res){
+	var userId = req.token.userId;
+	var ideaId = req.params.id;
+
+	var data = req.body;
+	var idea = null;
+	var newIdea = null;
+	var errMsg = null;
+
+	const STATE_CREATE_IDEA = 1;
```

Comment:
I don't see any benefit to employ state machine here. In your code, only one business logic branch for "create" or "update".  Overkill.  Again state machine is heavy and should be only for complex states. It's difficult to understand and maintain in the future.